### PR TITLE
Support for transforms / filters

### PR DESCRIPTION
**Pull request by david-zacharias**
In a recent project I have JSON-data that contains `\n` which I want partially (= when the screen designer decides its good to do) convert to `<br />`. So I needed a way for applying *transforms* (or may call it filters) in the template. This PR introduces a syntax like `{{myvar | transfrom}}` where `transform` is a callback that can be registered in the options. I have added documentation and tests. So please take a look and if you agree that this is a good feature, please accept the PR. I did not touch the existing tests, so I think it does not cause backward compatibiltiy issues.

### CODE REVIEW

File: tasks/bake.js [hunk 1]
```diff
@@ -25,7 +25,9 @@ module.exports = function( grunt ) {
 			section: null,
 			semanticIf: false,
 			basePath: "",
-			parsePattern: /\{\{\s*([\.\-\w]*)\s*\}\}/g
+			transforms: {},
+			parsePattern: /\{\{\s*([^\}]+)\s*\}\}/g,
+			transformGutter: '|'
```
Comment (MathiasPaumgarten) [inline tasks/bake.js]: please use double quote `"` as I use those in the rest of the project.

File: tasks/bake.js [hunk 2]
```diff
@@ -38,10 +40,19 @@ module.exports = function( grunt ) {
...
 
 		// This process method is used when no process function is supplied.
-
 		function defaultProcess( template, content ) {
-			return template.replace( options.parsePattern, function( match, key ) {
-				return resolveName( key, content );
+			return template.replace( options.parsePattern, function( match, inner ) {
+
+				// remove whitespace
+				var transforms = mout.array.map( inner.split( options.transformGutter ), function( str ) {
```
Comment (MathiasPaumgarten) [inline tasks/bake.js]: since we know we are in node, we can rely on native `map`

File: tasks/bake.js [hunk 3]
```diff
@@ -38,10 +40,19 @@ module.exports = function( grunt ) {
...
+					return mout.string.trim( str );
+				});
+
+				// the first value is our variable key and not a transfrom
+				var key = transforms.shift();
+				var resolved = resolveName( key, content );
+
+				return mout.array.reduce( transforms, applyTransform, resolved );
```
Comment (MathiasPaumgarten) [inline tasks/bake.js]: same as `map` we can use native `reduce`

File: README.md [hunk 4]
```diff
@@ -149,6 +149,30 @@ Determines the base directory for includes that are specified with an absolute p
 
 `<!--(bake /includes/footer.html)-->` relative to the basePath (level of Gruntfile by default)
 
+#### options.transforms
+Type: `Object`
+Default value: {}
+
+Registers callbacks that can be used as transforms in the tempalte with `{{myvar | upper}}`. It is possible to chain transforms like `{{myvar | upper | nl2br}}`.
```
Comment (MathiasPaumgarten) [inline README.md]: small typo in `template`

**Comment by MathiasPaumgarten** [comment]
I really like it. Additionally to some small code changes, I was wondering if we could merge some of the test cases in one or two test files. Currently this introduces 5 more test files, which is fine, but maybe we can narrow that down a bit?

**Comment by david-zacharias** [comment]
I am happy to hear that you like the idea. I fully agree with your remarks; code is updated. Regarding narrowing down the tests: Do you have an idea for reduction in your mind? I think it's better to have small tests which make it easy to track down a problem of failing tests.

**Comment by MathiasPaumgarten** [comment]
I see your point about the failing tests. Probably easier to see which one failed. Other than that I was gonna say we could put all files other than the deep transform into one test file. As they are not conflicting with each other. 

But I'm perfectly happy with this solution as well. Thank you for the PR. I'm gonna merge this now.

**Comment by MathiasPaumgarten** [comment]
Available under version `0.3.18`

**Pull request merged**